### ISSUE TITLE
TimSim Config Arguments

### ISSUE THREAD

**Issue by Philipp-Weigand** [theGreatHerrLebert/rustims#425]
Dear rustims team,

I can not find an overview of all possible arguments for the .toml config file used to simulate a measurement.
The files provided at the dda-PASEF HLA experiment (https://zenodo.org/uploads/15823046) already provide a good starting point, but I am missing a full list of available arguments.
Specifically, I am interested in the "annotation mode" that you mention in line 258 of the TimSim paper and the resulting segmentation ground truth. In your .toml files, I could not find any annotation mode argument.

Apologies, if I am missing any important documention.

Best,
Philipp

**Comment by theGreatHerrLebert** [comment]
Hi Philipp,

thanks, you are right that a full argument reference was missing (and there have been some additions since we posted the pre-print of timsim).

I have just added one. `CONFIGURATION.md` [here](https://github.com/theGreatHerrLebert/rustims/blob/main/packages/imspy-simulation/CONFIGURATION.md) now lists every TOML key together with its type, default value, and description, as well as the available CLI overrides and the instrument-specific configuration blocks.

Regarding **annotation mode** / segmentation ground truth, there are actually two distinct outputs:

1.  **Molecule-level ground truth.** Every simulation writes a
    `synthetic_data.db` file containing the complete molecular layout
    (peptides, ions, fragments, etc.).
2.  **Peak-level annotations.** The per-peak peptide/charge/isotope
    labels described in the paper are **not** written into the raw file
    and are **not** controlled by a TOML option. Instead, they are
    exposed through a Python API "streamer" intended for machine
    learning workflows, using:
    -   `TimsTofSyntheticPrecursorFrameBuilder.build_precursor_frame_annotated()`
    -   `to_dense_windows_with_labels()`

My read is that you are likely interested in the Peak-level annotation mode, for this one we added a notebook how it can be used for example for doing some experimental model training on raw data, have a look at the [training](https://github.com/theGreatHerrLebert/timsim-bench/blob/main/timsim_bench/notebooks/examples/AnnotatedFrameExample%20-%20Training.ipynb) and [inference](https://github.com/theGreatHerrLebert/timsim-bench/blob/main/timsim_bench/notebooks/examples/AnnotatedFrameExample%20-%20Inference.ipynb) notebooks.

Hope this helps and let me know if you run into any issues.

Best,
David

**Comment by Philipp-Weigand** [comment]
Dear David,

thank you for you quick and detailed response!
I am indeed interested in the Peak-level annotations. If I understand correctly, your methods provide m/z - Ion Mobility maps and corresponding Peak labels. This is already very helpful!
I am wondering, whether you also provide a method for m/z - Retention Time maps and labels, because m/z - RT maps are traditionally used for feature detection and existing Deep Learning approaches might benefit from your labels as well.
As far as I saw, I would need to stack the frames manually. Is that correct?

Best,
Philipp

**Comment by theGreatHerrLebert** [comment]
Hi Philipp,

You are right on both counts. The annotated API is intentionally per-frame, so to_dense_windows_with_labels() returns an m/z × ion-mobility grid (along with the isotopologue, charge, and feature-ID labels) for a single retention-time point. There is not currently a built-in annotated m/z–RT map or slice-level reader, so yes, you would stack the frames yourself along the RT dimension.

The good news is that this is relatively inexpensive. build_precursor_frames_annotated(frame_ids, num_threads=...) builds many annotated frames in a single multi-threaded call, and each TimsFrameAnnotated exposes both .frame_id and .retention_time. The workflow is therefore straightforward: build the annotated precursor frames, then for each frame either use to_dense_windows_with_labels() directly or marginalize the intensity and label arrays over the scan (ion-mobility) axis to obtain per-m/z-bin values. Finally, index the resulting arrays by .retention_time to assemble the m/z–RT heatmap together with the corresponding label maps.

Ah, and do not use the frame1 + frame2 operator for this... It merges peaks by (scan, quantized m/z) and averages the ion mobility, which destroys the retention-time information you want to preserve. Instead, stack the frames into a new array indexed by retention time.

An annotated m/z–RT view is a reasonable feature to provide natively, so could open a tracking issue for a helper that projects annotated frames onto the RT axis if that helps you. I am also happy to share a short example snippet if that would be helpful.

Best,
David

**Comment by Philipp-Weigand** [comment]
Hi David,

I would really appreciate a short example snippet!

Also, I struggle to create the intended m/z - ion-mobility map, because I currently struggle to understand the meaning behind the dimensions of the outputs of the `to_dense_windows_with_labels` function or its overall output purpose (I can't see its functionality). 
`scan_idex`, `window_index`, `mz` and `mobility` all have the shape (N, ), while `X`, `iso_labels`, `charge_labels` and `pep_labels` have the shape (N, Y). But what is N and Y here?
I would have expected that I can build an annotated m/z - ion-mobility map from these, but they often contain duplicate values.

I would be grateful for clarifications / examples on this.
By the way, I really like your support, its awesome!

Best,
Philipp

**Comment by theGreatHerrLebert** [comment]
Hi Philipp,

Happy to help, nice to see someone actually trying something with this feature!

Rather than just sending a snippet, I put together a complete onboarding notebook that walks through the annotated-frame API end-to-end and addresses each of your questions. It runs from top to bottom against any TimSim run; you only need to point DATASET_DB at a synthetic_data.db file:

https://github.com/theGreatHerrLebert/rustims/blob/main/packages/imspy-simulation/notebooks/AnnotatedFrames_Onboarding.ipynb

The short answers are:

What N and Y are. to_dense_windows_with_labels() does not return a single image. Instead, it returns a stack of fixed-width tiles, where each tile corresponds to one ion-mobility scan restricted to a single m/z window. Y (the columns) is fixed at round(window_length · 10^resolution) + 1. With the default settings (window_length = 10, resolution = 1), this gives 101 bins with a spacing of 0.1 Th. N (the rows) is the number of (scan × m/z-window) tiles that pass the min_num_peaks and min_intensity filters.

Why the duplicates appear. With overlapping=True, every m/z region is covered by two windows: one aligned and one shifted by half a window. Each scan therefore contributes one tile for every window that contains peaks, so the same scan, m/z, and mobility values can legitimately appear multiple times across different rows. The output is designed as a training-patch representation rather than a dense image.

Building the m/z-ion-mobility map. The notebook demonstrates two approaches. The first simply grids the raw peaks from frame.df. The second, which is closer to what you originally tried, reconstructs the full image directly from the dense-window output using absolute_bin = window_start(win_idx) + column. This reconstruction naturally resolves the overlap duplicates.

The notebook also covers building an m/z-retention-time map by stacking annotated frames along the RT axis (the feature-detection view you originally asked about), working with annotated fragment (MS2) frames, and includes a small PyTorch example. One note on the last section: it is intended only as an illustration of how the labels can be consumed, not as a recommended model architecture. For data this sparse, sparse or point-cloud models operating directly on the raw peaks are often a more natural choice. The labels simply provide ground truth that can be used for almost any supervised learning task.

Likely, one should spend some time thinking about the best raw data representation that works with machine learning algorithms that usually operate on dense data, if you are interested, check out how [casanovo dealt with it](https://www.nature.com/articles/s41467-024-49731-x).

Best,

David

**Comment by Philipp-Weigand** [comment]
Dear David,

Thanks for your help. I worked with your Onboarding notebook and realized, that it only takes a synthetic_data.db file as input and not the created .d file itself. Does this mean, that only synthetic data without any added noise or real data from a reference file is used in this example and neither the .tdf or .tdf_bin files are used?  If this is the case, it would explain why I don't see any reference signals. In that case, could you also please adapt the notebook so that all signals (synthetic and real onces) are utilized in the example?

Best,
Philipp

**Comment by theGreatHerrLebert** [comment]
Hi Philipp,

Good catch, and you are right on both counts, but I will give a bit more background so
it is clear how this all fits together.

## The two ground truths, and where the reference `.d` comes in

A TimSim run juggles two different representations, and the confusion here comes from
where the boundary between them sits.

`synthetic_data.db` holds the **molecular layout** in physical units: peptides, ions,
fragments, with m/z and inverse ion mobility. No TOF indices, no scan numbers, no
instrument calibration.

The written `.d` holds the **instrument view**: peaks placed on a real TOF/scan grid,
with Bruker calibration tables, an acquisition schedule and a frame layout.

The reference `.d` is what bridges the two, and it is doing four separate jobs:

1. **Calibration and axis conversion.** Its `MzCalibration`, `TimsCalibration` and
   `GlobalMetadata` tables are copied into your output, and every m/z to TOF and mobility
   to scan conversion at write time goes through it. `num_scans` comes from it too.
2. **Acquisition geometry.** The cycle time is derived from the reference frame table
   (which sets the frame count), and with `use_reference_layout` the DIA window layout is
   taken from it as well.
3. **Noise.** `add_real_data_noise` samples peaks from it. This is the job you are
   interested in.
4. It is **mandatory** for Bruker runs. Thermo and SCIEX swap it for a vendor template,
   Waters SONAR needs no file at all.

The key point is that all four of those happen on the **write** side. The annotated API
sits on the other side of that boundary: it re-derives frames from the database, which
already stores everything in physical units, so it needs no calibration and no reference.
`TimsTofSyntheticPrecursorFrameBuilder` therefore takes a database path and nothing else.
That is also why annotated frames come back with `tof` and `scan` set to zero.

## Why your frames are clean

Given that split, the pipeline order explains what you saw. TimSim builds the frames,
*then* adds noise in memory, *then* writes:

```
build frames  ->  add_real_data_noise_to_frames(...)  ->  tdf_writer.write_frame(...)
```

The noise peaks only ever reach the written `.tdf_bin`. They are never persisted back to
`synthetic_data.db`, so the annotated builder cannot see them. That is why the label
composition cell reports zero unannotated peaks, and why you see no reference signal.

The consequence worth being explicit about: the noise in your existing `.d` cannot be
recovered. It is sampled randomly per frame and never stored, so it can be *regenerated*
(statistically equivalent) but not *recovered* (the identical peaks). For training data
that is normally fine. If you need to reproduce one specific written frame peak for peak,
it is not.

## What you can do instead

Build frames that carry both signal labels and noise labels. The annotation model already
has a noise class:

* `SourceType` is one of `Signal`, `ChemicalNoise`, `RandomNoise`, `Unknown`.
* `TimsFrame.to_noise_annotated_tims_frame()` promotes a plain `TimsFrame` into a
  `TimsFrameAnnotated` in which every peak is labelled `RandomNoise`.
* `TimsFrameAnnotated` supports `+`.

So the recipe is a one liner:

```python
noisy = signal_frame + noise_frame.to_noise_annotated_tims_frame()
```

I have added a **section 10 ("Noise")** to the onboarding notebook that walks through this
end to end, with two variants for the `noise_frame`.

**Variant A, a synthetic background.** No reference file needed. You generate the peaks
yourself and tune density and brightness directly. This is the route I sketched for you in
 #440.

**Variant B, copied from a real reference `.d`.** This is what `add_real_data_noise = true`
does. The sampler lives on the reference dataset itself, so you can call it directly
without going through an acquisition builder:

```python
ref = TimsDatasetDIA(REFERENCE_D, in_memory=False, use_bruker_sdk=False)
noise = ref.sample_precursor_signal(num_frames=5, max_intensity=30.0, take_probability=0.2)
noisy = frame + noise.to_noise_annotated_tims_frame()
```

Those three arguments are exactly your `num_precursor_noise_frames`,
`reference_noise_intensity_max` and `precursor_sample_fraction`. This also gives you a fast
way to debug the problem from #440 in isolation: run just this sampler against your
reference and count the peaks it returns. If it comes back nearly empty, the reference has
nothing to copy in its MS1 frames, and no config value will change that.

## The noise label downstream

Noise flows through `to_dense_windows_with_labels` as its own label, so you can train
signal against background directly. The sentinels are:

| value | meaning |
|-------|---------|
| `-1`  | empty bin, or a source type that is neither signal nor random noise |
| `-2`  | `RandomNoise` |
| `0...5` | signal, a per-tile feature index, **not** a `peptide_id` |

That last row is worth flagging, since it surprised me too when re-reading the code. `pep`
is re-indexed within each tile and clamped at 5, so it answers "which distinct feature is
this bin part of, in this tile". It is a segmentation target, not a global identifier. Use
`frame.df.peptide_id` when you need the real peptide.

## Two caveats

The new section is deliberately demo only. A normal TimSim run still writes noise into the
`.d` alone, and nothing in `timsim/` emits noise-annotated frames today, so you compose
them yourself as shown.

Also, `+` merges peaks that land in the same `(scan, quantized m/z)` bin, and the
dense-window label then follows the strongest contributor. Bin counts therefore shift in
both directions and will not add up exactly.

To be precise about what is and is not missing here: TimSim does write noisy frames, the
noise is in your `.d`. What it does not do is keep any record of *which* peaks were noise.
The sampled peaks go into the `.tdf_bin` and nowhere else, so no labels can be attached to
them afterwards. If it would be useful to you, I can open a tracking issue for persisting
the sampled noise (or the per-frame sampler seed) into `synthetic_data.db`, which would let
the annotated builder reconstruct noisy frames with `RandomNoise` labels directly. That is
a pipeline change rather than a notebook one.

Best,
David
